File: ios/StemDeckRemote/Playback/PlayerViewModel.swift
import Foundation
import OSLog

@MainActor
final class PlayerViewModel: ObservableObject {
    private static let logger = Logger(subsystem: "com.stemdeck.remote", category: "PlayerViewModel")
    struct StemChannel: Identifiable, Equatable {
        let id: String
        var volume: Double = 1.0
        var isMuted: Bool = false
        var isSoloed: Bool = false

        var levelLabel: String {
            guard volume > 0.0001 else { return "-\u{221e} dB" }
            let db = 20 * log10(volume)
            return String(format: "%+.1f dB", db)
        }
    }

    enum State: Equatable {
        case waitingForStems
        case loadingEngine
        case ready
        case failed(String)
    }

    @Published private(set) var channels: [StemChannel] = []
    @Published private(set) var isPlaying = false
    @Published private(set) var state: State = .waitingForStems
    @Published private(set) var currentTime: TimeInterval = 0
    @Published private(set) var peaks: [String: [[Double]]] = [:]
    @Published private(set) var playbackRate: Double = 1.0
    @Published private(set) var pitchSemitones: Int = 0
    @Published private(set) var isLooping = false
    /// Live post-fader RMS level per stem, 0...1, updated while playing —
    /// what makes the LED meters bounce with the music instead of just
    /// showing the fader position. Rows fall back to the fader position
    /// whenever a stem has no recent entry here (paused, or before the
    /// first buffer arrives).
    @Published private(set) var liveLevels: [String: Double] = [:]
    @Published private(set) var currentChord: String?
    @Published private(set) var isAnalyzingChords = false
    private(set) var duration: TimeInterval = 0
    private var chordEvents: [ChordEvent] = []
    private var chordTask: Task<Void, Never>?

    static let speedRange = 0.5...1.5
    static let pitchRange = -6...6

    var progress: Double {
        guard duration > 0 else { return 0 }
        return min(max(currentTime / duration, 0), 1)
    }

    private let job: Job
    private let server: PairedServer
    private let store = StemFileStore()
    private let engine = StemMixerEngine()
    private let downloadQueue = StemDownloadQueue.shared
    private var progressTask: Task<Void, Never>?

    /// Non-nil only for a bundled `SampleSongCatalog` entry: stem name ->
    /// file URL straight in the app bundle. When set, `start()` skips the
    /// download queue and the server-backed peaks fetch entirely — nothing
    /// here ever touches the network.
    private let localStemURLs: [String: URL]?
    private let localPeaksURL: URL?

    init(job: Job, server: PairedServer, localStemURLs: [String: URL]? = nil, localPeaksURL: URL? = nil) {
        self.job = job
        self.server = server
        self.localStemURLs = localStemURLs
        self.localPeaksURL = localPeaksURL
    }

    func start() async {
        state = .waitingForStems
        guard !job.stemNames.isEmpty else {
            state = .failed("This song has no stems available.")
            return
        }

        let stemURLs: [String: URL]
        if let localStemURLs {
            stemURLs = localStemURLs
            loadLocalPeaksIfNeeded()
        } else {
            do {
                try await downloadQueue.downloadNow(job)
            } catch {
                if case .failed(let message) = downloadQueue.status(for: job.id) {
                    state = .failed(message)
                } else {
                    state = .failed("Couldn't download stems.")
                }
                return
            }
            await loadPeaksIfNeeded()
            stemURLs = Dictionary(uniqueKeysWithValues: job.stemNames.map { ($0, store.url(jobID: job.id, stem: $0)) })
        }

        state = .loadingEngine

        do {
            let stems = job.stemNames.compactMap { name in stemURLs[name].map { (name: name, url: $0) } }
            try engine.load(stems: stems) { [weak self] in
                self?.isPlaying = false
                self?.currentTime = 0
                self?.liveLevels.removeAll()
            }
            engine.onStemLevel = { [weak self] stem, rms in
                self?.updateLiveLevel(stem: stem, rms: rms)
            }
            duration = engine.duration
            channels = job.stemNames.map { StemChannel(id: $0) }
            state = .ready
            startProgressUpdates()
        } catch {
            state = .failed("Couldn't load stems for playback: \(error.localizedDescription)")
        }
    }

    private func loadPeaksIfNeeded() async {
        let jobID = job.id
        if !store.peaksExist(jobID: jobID) {
            let client = StemDeckClient(server: server)
            if let data = try? await client.fetchPeaks(jobID: jobID) {
                try? data.write(to: store.peaksURL(jobID: jobID))
            }
        }
        guard let data = try? Data(contentsOf: store.peaksURL(jobID: jobID)) else { return }
        peaks = (try? JSONDecoder().decode([String: [[Double]]].self, from: data)) ?? [:]
    }

    private func loadLocalPeaksIfNeeded() {
        guard let localPeaksURL, let data = try? Data(contentsOf: localPeaksURL) else { return }
        peaks = (try? JSONDecoder().decode([String: [[Double]]].self, from: data)) ?? [:]
    }

    func togglePlayback() {
        if isPlaying {
            engine.pause()
            liveLevels.removeAll()
        } else {
            engine.play()
        }
        isPlaying = engine.isPlaying
    }

    /// RMS of real program material almost never reaches 0dBFS — that only
    /// happens for a constant full-scale wave, effectively clipping — so
    /// mapping all the way up to 0dB (as this used to) meant the top LEDs
    /// were essentially unreachable even at a 100% fader. -6dB is a much
    /// more realistic "meter's full" reference for loud passages.
    private static let meterFloorDb: Double = -40
    private static let meterCeilingDb: Double = -6

    private func updateLiveLevel(stem: String, rms: Float) {
        let db = Double(20 * log10(max(rms, 0.0001)))
        let normalized = (db - Self.meterFloorDb) / (Self.meterCeilingDb - Self.meterFloorDb)
        liveLevels[stem] = min(max(normalized, 0), 1)
    }

    func seek(to time: TimeInterval) {
        engine.seek(to: time)
        currentTime = engine.currentTime()
        isPlaying = engine.isPlaying
    }

    func setVolume(_ volume: Double, for stem: String) {
        engine.setVolume(Float(volume), for: stem)
        if let index = channels.firstIndex(where: { $0.id == stem }) {
            channels[index].volume = volume
        }
    }

    func toggleMute(_ stem: String) {
        engine.toggleMute(stem)
        if let index = channels.firstIndex(where: { $0.id == stem }) {
            channels[index].isMuted = engine.isMuted(stem)
        }
    }

    func toggleSolo(_ stem: String) {
        engine.toggleSolo(stem)
        for index in channels.indices {
            channels[index].isSoloed = engine.isSoloed(channels[index].id)
        }
    }

    func setPlaybackRate(_ rate: Double) {
        let clamped = min(max(Self.speedRange.lowerBound, rate), Self.speedRange.upperBound)
        engine.setRate(Float(clamped))
        playbackRate = clamped
    }

    func setPitch(_ value: Double) {
        let rounded = Int(value.rounded())
        let clamped = min(max(Self.pitchRange.lowerBound, rounded), Self.pitchRange.upperBound)
        guard clamped != pitchSemitones else { return }
        engine.setPitch(semitones: clamped)
        pitchSemitones = clamped
    }

    func toggleLoop() {
        isLooping.toggle()
        engine.isLooping = isLooping
    }

    func stopPlayback() {
        engine.pause()
        engine.seek(to: 0)
        isPlaying = false
        currentTime = 0
        liveLevels.removeAll()
    }

    func markIn() {}
    func markOut() {}

    private func startProgressUpdates() {
        progressTask?.cancel()
        progressTask = Task { [weak self] in
            while !Task.isCancelled {
                guard let self else { return }
                self.currentTime = self.engine.currentTime()
                self.updateCurrentChord()
                try? await Task.sleep(nanoseconds: 50_000_000)
            }
        }
    }

    private func updateCurrentChord() {
        currentChord = chordEvents.first { $0.start <= currentTime && currentTime < $0.end }?.chord
    }

    func stop() {
        progressTask?.cancel()
        chordTask?.cancel()
        isAnalyzingChords = false
        engine.stopAndReset()
        isPlaying = false
    }
}

File: android/app/src/main/kotlin/com/stemdeck/remote/playback/PlayerViewModel.kt
package com.stemdeck.remote.playback

import android.content.Context
import android.net.Uri
import android.util.Log
import com.stemdeck.remote.models.Job
import com.stemdeck.remote.models.PairedServer
import com.stemdeck.remote.networking.StemDeckClient
import kotlin.math.log10
import kotlin.math.max
import kotlin.math.min
import kotlinx.coroutines.CoroutineScope
import kotlinx.coroutines.Dispatchers
import kotlinx.coroutines.Job as CoroutineJob
import kotlinx.coroutines.SupervisorJob
import kotlinx.coroutines.delay
import kotlinx.coroutines.flow.MutableStateFlow
import kotlinx.coroutines.flow.StateFlow
import kotlinx.coroutines.flow.update
import kotlinx.coroutines.isActive
import kotlinx.coroutines.launch
import kotlinx.serialization.json.Json

/**
 * Drives one song's stem console: downloads (or reads bundled) stems, loads
 * them into a [StemMixerEngine], and exposes everything the UI needs as
 * [StateFlow]s — the direct Android counterpart of the iOS side's
 * `PlayerViewModel` (`ObservableObject` + `@Published`).
 *
 * Owned by [PlaybackCoordinator], not by any single screen — the console can
 * be minimized to the mini player without stopping playback, exactly like
 * the iOS side.
 */
class PlayerViewModel(
    private val context: Context,
    private val job: Job,
    private val server: PairedServer,
    private val stemDownloadQueue: StemDownloadQueue,
    private val stemFileStore: StemFileStore,
    /** Non-null only for a bundled [com.stemdeck.remote.pairing.SampleSongCatalog] entry: stem name -> `asset:///...` URI. When set, [start] skips the download queue and the server-backed peaks fetch entirely — nothing here ever touches the network. */
    private val localStemUris: Map<String, Uri>? = null,
    private val localPeaksAssetPath: String? = null,
) {
    data class StemChannel(
        val id: String,
        val volume: Double = 1.0,
        val isMuted: Boolean = false,
        val isSoloed: Boolean = false,
    ) {
        val levelLabel: String
            get() {
                if (volume <= 0.0001) return "-∞ dB"
                val db = 20 * log10(volume)
                return "%+.1f dB".format(db)
            }
    }

    sealed class State {
        object WaitingForStems : State()
        object LoadingEngine : State()
        object Ready : State()
        data class Failed(val message: String) : State()
    }

    private val _channels = MutableStateFlow<List<StemChannel>>(emptyList())
    val channels: StateFlow<List<StemChannel>> get() = _channels

    private val _isPlaying = MutableStateFlow(false)
    val isPlaying: StateFlow<Boolean> get() = _isPlaying

    private val _state = MutableStateFlow<State>(State.WaitingForStems)
    val state: StateFlow<State> get() = _state

    private val _currentTime = MutableStateFlow(0.0)
    val currentTime: StateFlow<Double> get() = _currentTime

    private val _peaks = MutableStateFlow<Map<String, List<List<Double>>>>(emptyMap())
    val peaks: StateFlow<Map<String, List<List<Double>>>> get() = _peaks

    private val _playbackRate = MutableStateFlow(1.0)
    val playbackRate: StateFlow<Double> get() = _playbackRate

    private val _pitchSemitones = MutableStateFlow(0)
    val pitchSemitones: StateFlow<Int> get() = _pitchSemitones

    private val _isLooping = MutableStateFlow(false)
    val isLooping: StateFlow<Boolean> get() = _isLooping

    /** Live post-fader RMS level per stem, 0..1, updated while playing — what makes the LED meters bounce with the music instead of just showing the fader position. */
    private val _liveLevels = MutableStateFlow<Map<String, Double>>(emptyMap())
    val liveLevels: StateFlow<Map<String, Double>> get() = _liveLevels

    /** Never actually populated today — see [ChordAnalyzer]'s doc comment: chord detection is fully ported but not wired into playback yet, matching the iOS app's current (unwired) state. */
    private val _currentChord = MutableStateFlow<String?>(null)
    val currentChord: StateFlow<String?> get() = _currentChord

    private val _isAnalyzingChords = MutableStateFlow(false)
    val isAnalyzingChords: StateFlow<Boolean> get() = _isAnalyzingChords

    private val _duration = MutableStateFlow(0.0)
    val duration: StateFlow<Double> get() = _duration

    val progress: Double
        get() {
            val d = _duration.value
            if (d <= 0) return 0.0
            return (_currentTime.value / d).coerceIn(0.0, 1.0)
        }

    private val engine = StemMixerEngine(context)
    private val scope = CoroutineScope(SupervisorJob() + Dispatchers.Main.immediate)
    private var progressJob: CoroutineJob? = null

    suspend fun start() {
        _state.value = State.WaitingForStems
        if (job.stemNames.isEmpty()) {
            _state.value = State.Failed("This song has no stems available.")
            return
        }

        if (localStemUris != null) {
            loadLocalPeaksIfNeeded()
        } else {
            try {
                stemDownloadQueue.downloadNow(job)
            } catch (e: Exception) {
                val status = stemDownloadQueue.status(job.id)
                _state.value = State.Failed(
                    if (status is StemDownloadQueue.Status.Failed) status.message else "Couldn't download stems.",
                )
                return
            }
            loadPeaksIfNeeded()
        }

        _state.value = State.LoadingEngine

        try {
            val stems = job.stemNames.map { name ->
                val uri = localStemUris?.get(name) ?: Uri.fromFile(stemFileStore.url(job.id, name))
                StemMixerEngine.Stem(name, uri)
            }
            engine.onStemLevel = { stem, rms -> updateLiveLevel(stem, rms) }
            engine.load(stems) {
                _isPlaying.value = false
                _currentTime.value = 0.0
                _liveLevels.value = emptyMap()
            }
            _duration.value = engine.durationMs / 1000.0
            _channels.value = job.stemNames.map { StemChannel(id = it) }
            _state.value = State.Ready
            startProgressUpdates()
        } catch (e: Exception) {
            _state.value = State.Failed("Couldn't load stems for playback: ${e.message}")
        }
    }

    private suspend fun loadPeaksIfNeeded() {
        val jobID = job.id
        if (!stemFileStore.peaksExist(jobID)) {
            try {
                val client = StemDeckClient(server)
                val data = client.fetchPeaks(jobID)
                stemFileStore.peaksFile(jobID).writeBytes(data)
            } catch (e: Exception) {
                // Peaks are optional — WaveformView just draws nothing without them.
            }
        }
        try {
            val text = stemFileStore.peaksFile(jobID).readText()
            _peaks.value = Json.decodeFromString(text)
        } catch (e: Exception) {
            // Missing/corrupt peaks file — leave peaks empty.
        }
    }

    private fun loadLocalPeaksIfNeeded() {
        val path = localPeaksAssetPath ?: return
        try {
            val text = context.assets.open(path).bufferedReader().use { it.readText() }
            _peaks.value = Json.decodeFromString(text)
        } catch (e: Exception) {
            // Missing/corrupt bundled peaks — leave peaks empty.
        }
    }

    fun togglePlayback() {
        if (_isPlaying.value) {
            engine.pause()
            _liveLevels.value = emptyMap()
        } else {
            engine.play()
        }
        _isPlaying.value = engine.isPlaying
    }

    /** RMS of real program material almost never reaches 0dBFS, so mapping all the way up to 0dB would leave the top LEDs essentially unreachable even at a 100% fader — -6dB is a much more realistic "meter's full" reference for loud passages. */
    private fun updateLiveLevel(stem: String, rms: Float) {
        val db = 20 * log10(max(rms, 0.0001f).toDouble())
        val normalized = (db - METER_FLOOR_DB) / (METER_CEILING_DB - METER_FLOOR_DB)
        _liveLevels.update { it + (stem to normalized.coerceIn(0.0, 1.0)) }
    }

    fun seek(timeSeconds: Double) {
        engine.seek((timeSeconds * 1000).toLong())
        _currentTime.value = engine.currentPositionMs() / 1000.0
        _isPlaying.value = engine.isPlaying
    }

    fun setVolume(volume: Double, stem: String) {
        engine.setVolume(stem, volume.toFloat())
        _channels.update { list -> list.map { if (it.id == stem) it.copy(volume = volume) else it } }
    }

    fun toggleMute(stem: String) {
        engine.toggleMute(stem)
        _channels.update { list -> list.map { if (it.id == stem) it.copy(isMuted = engine.isMuted(stem)) else it } }
    }

    fun toggleSolo(stem: String) {
        engine.toggleSolo(stem)
        _channels.update { list -> list.map { it.copy(isSoloed = engine.isSoloed(it.id)) } }
    }

    fun setPlaybackRate(rate: Double) {
        val clamped = rate.coerceIn(SPEED_RANGE.first, SPEED_RANGE.second)
        engine.setRate(clamped.toFloat())
        _playbackRate.value = clamped
    }

    fun setPitch(value: Double) {
        val clamped = Math.round(value).toInt().coerceIn(PITCH_RANGE.first, PITCH_RANGE.last)
        if (clamped == _pitchSemitones.value) return
        engine.setPitchSemitones(clamped)
        _pitchSemitones.value = clamped
    }

    fun toggleLoop() {
        _isLooping.value = !_isLooping.value
        engine.isLooping = _isLooping.value
    }

    fun stopPlayback() {
        engine.pause()
        engine.seek(0)
        _isPlaying.value = false
        _currentTime.value = 0.0
        _liveLevels.value = emptyMap()
    }

    fun markIn() = Unit
    fun markOut() = Unit

    private fun startProgressUpdates() {
        progressJob?.cancel()
        progressJob = scope.launch {
            while (isActive) {
                _currentTime.value = engine.currentPositionMs() / 1000.0
                updateCurrentChord()
                delay(50)
            }
        }
    }

    private fun updateCurrentChord() {
        // chordEvents is always empty today — see the currentChord doc comment above.
    }

    fun stop() {
        progressJob?.cancel()
        _isAnalyzingChords.value = false
        engine.stopAndReset()
        _isPlaying.value = false
    }

    companion object {
        val SPEED_RANGE = 0.5 to 1.5
        val PITCH_RANGE = -6..6
        private const val METER_FLOOR_DB = -40.0
        private const val METER_CEILING_DB = -6.0
    }
}
